File: VB/PerformanceCounterCorrelationExample/PerformanceDataProvider.vb
Imports System
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Windows.Forms

Namespace PerformanceCounterCorrelation
    Public Class PerformanceDataProvider
        Private timerToCollectData As Timer
        Private logList As BindingList(Of TwoCounterSample)
        Private firstCounter As PerformanceCounter
        Private secondCounter As PerformanceCounter

        Public Sub New(ByVal first As PerformanceCounter, ByVal second As PerformanceCounter)
            timerToCollectData = New Timer()
            logList = New BindingList(Of TwoCounterSample)()
            timerToCollectData.Interval = 500
            firstCounter = first
            secondCounter = second
            AddHandler timerToCollectData.Tick, AddressOf TimerToCollectData_Tick
        End Sub


        Private Sub TimerToCollectData_Tick(ByVal sender As Object, ByVal e As EventArgs)
            LogPerformanceData()
        End Sub

        Private Sub LogPerformanceData()
            If logList.Count > 500 Then
                [Stop]()
                RaiseDataOver(New EventArgs())
            End If
            logList.Add(New TwoCounterSample(firstCounter.NextValue(), secondCounter.NextValue()))
            If logList.Count = 3 Then
                RaiseDataReady(New EventArgs())
            End If
        End Sub

        Public Sub Start(ByVal interval As Integer)
            timerToCollectData.Interval = interval
            timerToCollectData.Enabled = True
        End Sub

        Private Event onDataReady As EventHandler
        Public Custom Event DataReady As EventHandler
            AddHandler(ByVal value As EventHandler)
                AddHandler onDataReady, value
            End AddHandler
            RemoveHandler(ByVal value As EventHandler)
                RemoveHandler onDataReady, value
            End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As System.EventArgs)
                RaiseEvent onDataReady(sender, e)
            End RaiseEvent
        End Event
        Private Sub RaiseDataReady(ByVal e As EventArgs)
            RaiseEvent onDataReady(Me,e)
        End Sub

        Public Function GetData() As IBindingList
            Return logList
        End Function

        Private Event onDataOver As EventHandler
        Public Custom Event DataOver As EventHandler
            AddHandler(ByVal value As EventHandler)
                AddHandler onDataOver, value
            End AddHandler
            RemoveHandler(ByVal value As EventHandler)
                RemoveHandler onDataOver, value
            End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As System.EventArgs)
                RaiseEvent onDataOver(sender, e)
            End RaiseEvent
        End Event
        Private Sub RaiseDataOver(ByVal e As EventArgs)
            RaiseEvent onDataOver(Me, e)
        End Sub

        Public Sub [Stop]()
            timerToCollectData.Enabled = False
        End Sub
    End Class
End Namespace

File: CS/PerformanceCounterCorrelationExample/PerformanceDataProvider.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Windows.Forms;

namespace PerformanceCounterCorrelation {
    public class PerformanceDataProvider {
        Timer timerToCollectData;
        BindingList<TwoCounterSample> logList;
        PerformanceCounter firstCounter;
        PerformanceCounter secondCounter;

        public PerformanceDataProvider(PerformanceCounter first, PerformanceCounter second) {
            timerToCollectData = new Timer();
            logList = new BindingList<TwoCounterSample>();
            timerToCollectData.Interval = 500;
            firstCounter = first;
            secondCounter = second;
            timerToCollectData.Tick += TimerToCollectData_Tick;
        }


        private void TimerToCollectData_Tick(object sender, EventArgs e) {
            LogPerformanceData();
        }

        private void LogPerformanceData() {
            if (logList.Count > 500) {
                Stop();
                RaiseDataOver(new EventArgs());
            }
            logList.Add(new TwoCounterSample(firstCounter.NextValue(), secondCounter.NextValue()));
            if (logList.Count == 3) RaiseDataReady(new EventArgs());
        }

        public void Start(int interval) {
            timerToCollectData.Interval = interval;
            timerToCollectData.Enabled = true;
        }

        event EventHandler onDataReady;
        public event EventHandler DataReady { add { onDataReady += value; } remove { onDataReady -= value; } }
        void RaiseDataReady(EventArgs e) {
            if (onDataReady != null) {
                onDataReady(this,e);
            }
        }

        public IBindingList GetData() {
            return logList;
        }

        event EventHandler onDataOver;
        public event EventHandler DataOver { add { onDataOver += value; } remove { onDataOver -= value; } }
        void RaiseDataOver(EventArgs e) {
            if (onDataOver != null) {
                onDataOver(this, e);
            }
        }

        public void Stop() {
            timerToCollectData.Enabled = false;
        }
    }
}
